File: app/src/main/java/com/vondear/tools/adapter/AdapterRightDish.java
package com.vondear.tools.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.vondear.tools.R;
import com.vondear.tools.model.ModelDish;
import com.vondear.tools.model.ModelDishMenu;
import com.vondear.tools.model.ModelShopCart;
import com.vondear.tools.interfaces.ShopCartInterface;

import java.util.ArrayList;

/**
 * @author vondear
 * @date 16-11-10
 */
public class AdapterRightDish extends RecyclerView.Adapter {
    private final int MENU_TYPE = 0;
    private final int DISH_TYPE = 1;
    private final int HEAD_TYPE = 2;

    private Context mContext;
    private ArrayList<ModelDishMenu> mMenuList;
    private int mItemCount;
    private ModelShopCart mModelShopCart;
    private ShopCartInterface shopCartImp;

    public AdapterRightDish(Context mContext, ArrayList<ModelDishMenu> mMenuList, ModelShopCart modelShopCart){
        this.mContext = mContext;
        this.mMenuList = mMenuList;
        this.mItemCount = mMenuList.size();
        this.mModelShopCart = modelShopCart;
        for(ModelDishMenu menu:mMenuList){
            mItemCount+=menu.getModelDishList().size();
        }
    }

    @Override
    public int getItemViewType(int position) {
        int sum=0;
        for(ModelDishMenu menu:mMenuList){
            if(position==sum){
                return MENU_TYPE;
            }
            sum+=menu.getModelDishList().size()+1;
        }
        return DISH_TYPE;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if(viewType==MENU_TYPE){
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.right_menu_item, parent, false);
            MenuViewHolder viewHolder = new MenuViewHolder(view);
            return viewHolder;
        }else{
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.right_dish_item1, parent, false);
            DishViewHolder viewHolder = new DishViewHolder(view);
            return viewHolder;
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, final int position) {
        if(getItemViewType(position)==MENU_TYPE){
            MenuViewHolder menuholder = (MenuViewHolder)holder;
            if(menuholder!=null) {
                menuholder.right_menu_title.setText(getMenuByPosition(position).getMenuName());
                menuholder.right_menu_layout.setContentDescription(position+"");
            }
        }else {
            final DishViewHolder dishholder = (DishViewHolder) holder;
            if (dishholder != null) {

                final ModelDish modelDish = getDishByPosition(position);
                dishholder.right_dish_name_tv.setText(modelDish.getDishName());
                dishholder.right_dish_price_tv.setText(modelDish.getDishPrice()+"");
                dishholder.right_dish_layout.setContentDescription(position+"");

                int count = 0;
                if(mModelShopCart.getShoppingSingleMap().containsKey(modelDish)){
                    count = mModelShopCart.getShoppingSingleMap().get(modelDish);
                }
                if(count<=0){
                    dishholder.right_dish_remove_iv.setVisibility(View.GONE);
                    dishholder.right_dish_account_tv.setVisibility(View.GONE);
                }else {
                    dishholder.right_dish_remove_iv.setVisibility(View.VISIBLE);
                    dishholder.right_dish_account_tv.setVisibility(View.VISIBLE);
                    dishholder.right_dish_account_tv.setText(count+"");
                }
                dishholder.right_dish_add_iv.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        if(mModelShopCart.addShoppingSingle(modelDish)) {
                            notifyItemChanged(position);
                            if(shopCartImp!=null)
                                shopCartImp.add(view,position);
                        }
                    }
                });

                dishholder.right_dish_remove_iv.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        if(mModelShopCart.subShoppingSingle(modelDish)){
                            notifyItemChanged(position);
                            if(shopCartImp!=null)
                                shopCartImp.remove(view,position);
                        }
                    }
                });
            }
        }
    }

    public ModelDishMenu getMenuByPosition(int position){
        int sum =0;
        for(ModelDishMenu menu:mMenuList){
            if(position==sum){
                return menu;
            }
            sum+=menu.getModelDishList().size()+1;
        }
        return null;
    }

    public ModelDish getDishByPosition(int position){
        for(ModelDishMenu menu:mMenuList){
            if(position>0 && position<=menu.getModelDishList().size()){
                return menu.getModelDishList().get(position-1);
            }
            else{
                position-=menu.getModelDishList().size()+1;
            }
        }
        return null;
    }

    public ModelDishMenu getMenuOfMenuByPosition(int position){
        for(ModelDishMenu menu:mMenuList){
            if(position==0)return menu;
            if(position>0 && position<=menu.getModelDishList().size()){
                return menu;
            }
            else{
                position-=menu.getModelDishList().size()+1;
            }
        }
        return null;
    }

    @Override
    public int getItemCount() {
        return mItemCount;
    }

    public ShopCartInterface getShopCartInterface() {
        return shopCartImp;
    }

    public void setShopCartInterface(ShopCartInterface shopCartImp) {
        this.shopCartImp = shopCartImp;
    }

    private class MenuViewHolder extends RecyclerView.ViewHolder{
        private LinearLayout right_menu_layout;
        private TextView right_menu_title;
        public MenuViewHolder(View itemView) {
            super(itemView);
            right_menu_layout = (LinearLayout)itemView.findViewById(R.id.right_menu_item);
            right_menu_title = (TextView)itemView.findViewById(R.id.right_menu_tv);
        }
    }

    private class DishViewHolder extends RecyclerView.ViewHolder{
        private TextView right_dish_name_tv;
        private TextView right_dish_price_tv;
        private LinearLayout right_dish_layout;
        private ImageView right_dish_remove_iv;
        private ImageView right_dish_add_iv;
        private TextView right_dish_account_tv;

        public DishViewHolder(View itemView) {
            super(itemView);
            right_dish_name_tv = (TextView)itemView.findViewById(R.id.right_dish_name);
            right_dish_price_tv = (TextView)itemView.findViewById(R.id.right_dish_price);
            right_dish_layout = (LinearLayout)itemView.findViewById(R.id.right_dish_item);
            right_dish_remove_iv = (ImageView)itemView.findViewById(R.id.right_dish_remove);
            right_dish_add_iv = (ImageView)itemView.findViewById(R.id.right_dish_add);
            right_dish_account_tv = (TextView) itemView.findViewById(R.id.right_dish_account);
        }

    }
}

File: RxDemo/src/main/java/com/tamsiree/rxdemo/adapter/AdapterRightDish.kt
package com.tamsiree.rxdemo.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.tamsiree.rxdemo.R
import com.tamsiree.rxdemo.interfaces.ShopCartInterface
import com.tamsiree.rxdemo.model.ModelDish
import com.tamsiree.rxdemo.model.ModelDishMenu
import com.tamsiree.rxdemo.model.ModelShopCart
import java.util.*

/**
 * @author tamsiree
 * @date 16-11-10
 */
class AdapterRightDish(private val mContext: Context, private val mMenuList: ArrayList<ModelDishMenu>, modelShopCart: ModelShopCart) : RecyclerView.Adapter<RecyclerView.ViewHolder?>() {
    private val MENU_TYPE = 0
    private val DISH_TYPE = 1
    private val HEAD_TYPE = 2
    private var mItemCount: Int
    private val mModelShopCart: ModelShopCart
    var shopCartInterface: ShopCartInterface? = null
    override fun getItemViewType(position: Int): Int {
        var sum = 0
        for (menu in mMenuList) {
            if (position == sum) {
                return MENU_TYPE
            }
            sum += menu.modelDishList!!.size + 1
        }
        return DISH_TYPE
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == MENU_TYPE) {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.right_menu_item, parent, false)
            MenuViewHolder(view)
        } else {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.right_dish_item1, parent, false)
            DishViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (getItemViewType(position) == MENU_TYPE) {
            val menuholder = holder as MenuViewHolder?
            if (menuholder != null) {
                menuholder.right_menu_title.text = getMenuByPosition(position)!!.menuName
                menuholder.right_menu_layout.contentDescription = position.toString() + ""
            }
        } else {
            val dishholder = holder as DishViewHolder?
            if (dishholder != null) {
                val modelDish = getDishByPosition(position)
                dishholder.right_dish_name_tv.text = modelDish!!.dishName
                dishholder.right_dish_price_tv.text = modelDish.dishPrice.toString() + ""
                dishholder.right_dish_layout.contentDescription = position.toString() + ""
                var count = 0
                if (mModelShopCart.shoppingSingleMap.containsKey(modelDish)) {
                    count = mModelShopCart.shoppingSingleMap[modelDish]!!
                }
                if (count <= 0) {
                    dishholder.right_dish_remove_iv.visibility = View.GONE
                    dishholder.right_dish_account_tv.visibility = View.GONE
                } else {
                    dishholder.right_dish_remove_iv.visibility = View.VISIBLE
                    dishholder.right_dish_account_tv.visibility = View.VISIBLE
                    dishholder.right_dish_account_tv.text = count.toString() + ""
                }
                dishholder.right_dish_add_iv.setOnClickListener(object : View.OnClickListener {
                    override fun onClick(view: View) {
                        if (mModelShopCart.addShoppingSingle(modelDish)) {
                            notifyItemChanged(position)
                            if (shopCartInterface != null) {
                                shopCartInterface!!.add(view, position)
                            }
                        }
                    }
                })
                dishholder.right_dish_remove_iv.setOnClickListener(object : View.OnClickListener {
                    override fun onClick(view: View) {
                        if (mModelShopCart.subShoppingSingle(modelDish)) {
                            notifyItemChanged(position)
                            if (shopCartInterface != null) {
                                shopCartInterface!!.remove(view, position)
                            }
                        }
                    }
                })
            }
        }
    }

    fun getMenuByPosition(position: Int): ModelDishMenu? {
        var sum = 0
        for (menu in mMenuList) {
            if (position == sum) {
                return menu
            }
            sum += menu.modelDishList!!.size + 1
        }
        return null
    }

    fun getDishByPosition(position: Int): ModelDish? {
        var position = position
        for (menu in mMenuList) {
            position -= if (position > 0 && position <= menu.modelDishList!!.size) {
                return menu.modelDishList!![position - 1]
            } else {
                menu.modelDishList!!.size + 1
            }
        }
        return null
    }

    fun getMenuOfMenuByPosition(position: Int): ModelDishMenu? {
        var position = position
        for (menu in mMenuList) {
            if (position == 0) {
                return menu
            }
            position -= if (position > 0 && position <= menu.modelDishList!!.size) {
                return menu
            } else {
                menu.modelDishList!!.size + 1
            }
        }
        return null
    }

    override fun getItemCount(): Int {
        return mItemCount
    }

    private inner class MenuViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val right_menu_layout: LinearLayout
        val right_menu_title: TextView

        init {
            right_menu_layout = itemView.findViewById<View>(R.id.right_menu_item) as LinearLayout
            right_menu_title = itemView.findViewById<View>(R.id.right_menu_tv) as TextView
        }
    }

    private inner class DishViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val right_dish_name_tv: TextView
        val right_dish_price_tv: TextView
        val right_dish_layout: LinearLayout
        val right_dish_remove_iv: ImageView
        val right_dish_add_iv: ImageView
        val right_dish_account_tv: TextView

        init {
            right_dish_name_tv = itemView.findViewById<View>(R.id.right_dish_name) as TextView
            right_dish_price_tv = itemView.findViewById<View>(R.id.right_dish_price) as TextView
            right_dish_layout = itemView.findViewById<View>(R.id.right_dish_item) as LinearLayout
            right_dish_remove_iv = itemView.findViewById<View>(R.id.right_dish_remove) as ImageView
            right_dish_add_iv = itemView.findViewById<View>(R.id.right_dish_add) as ImageView
            right_dish_account_tv = itemView.findViewById<View>(R.id.right_dish_account) as TextView
        }
    }

    init {
        mItemCount = mMenuList.size
        mModelShopCart = modelShopCart
        for (menu in mMenuList) {
            mItemCount += menu.modelDishList!!.size
        }
    }


}